### ISSUE TITLE
[Feature Request]: Show Pipeline Run pod logs per selected node

### ISSUE THREAD

**Issue by andrewballantyne** [opendatahub-io/odh-dashboard#1247]
### Feature description

Each node in the Pipeline Topology is backed by a concept called a Task. Each Task is a container and thus has it's own pod.

Today we show the pod id in the header of the tab selection -- we need a tab to show the pod logs.

This might require us to look at the OpenShift Console pod log component. We will need some idea of what features the log should have. Can it pause? We won't be able to "stream it" as we don't have web sockets -- but we could "fake stream it" by just delay adding lines to the log one after each other instead of bulk adding it.

This needs some UX to help understand how we show the details of the pod log.

### Anything else?

@kywalker-rh  this needs some UX attention before we can work on it.

**Comment by gregsheremeta** [comment]
Is there a rough ETA for this one?

**Comment by andrewballantyne** [comment]
> Is there a rough ETA for this one?

Not currently -- but I can look to get this into next sprint. @kywalker-rh we should talk about any UX left dangling, I'll send you a meeting invite.

**Comment by andrewballantyne** [comment]
Some implementation details:

* We will use the PF Extension, Log Viewer - https://www.patternfly.org/v4/extensions/log-viewer
    * NPM => https://www.npmjs.com/package/@patternfly/react-log-viewer?activeTab=versions
* Features we want to support for the first pass
    * Logs are done by container inside the pod -- we will use a dropdown to show all containers, select the first container
    * Tail the log file by 500 lines every 3 seconds -- don't cache any information, just show the current selected container
    * On each fetch we will toss the previous data away and just show the latest information (so the last - up to - 500 lines)
    * Put a message above the log viewer that specifies it is only showing the last 500 lines, use the download feature to view the entire log
    * Support a download button to pull all logs at once
        * Probably download per container -- TBD
* Web Sockets are needed to make this a real log viewer - out of scope - https://github.com/opendatahub-io/odh-dashboard/issues/1138

@kywalker-rh I think we are going to need a mock to help out the initial implementation. Best to just do up something quick and not worry too much about wording or small details, we can work that out on the PR & in incubation.

**Comment by yih-wang** [comment]
Hi @andrewballantyne and @kywalker-rh, [here is the mockup](https://www.figma.com/proto/rJjOtl3nxpW6BAcl0MZr4y/Pipeline-log-viewer?page-id=62%3A1895&type=design&node-id=82-1161&viewport=723%2C-184%2C0.24&t=S2A8z2MwR9c9Ffjg-1&scaling=min-zoom&starting-point-node-id=82%3A1161&mode=design) of this requirement.  

Regarding the features we want to support for the first pass, I have several thoughts/questions:
1. I have incorporated a functionality inspired by Openshift in my design, which involves abridging exceptionally long log lines (log lines that exceed 1021 characters in length). This is particularly important in our use case, where the log viewer is embedded in a drawer and refreshes every 3 seconds. By shortening these long log lines, we ensure a more user-friendly experience and prevent the logs from occupying excessive space on the screen.
2. I wonder whether we could also provide the pause/resume feature for the first pass. Imagine a scenario where the user is viewing specific log lines, the log viewer refreshes after 3 seconds, then the log the user was looking at suddenly disappeared. In such cases, I would assume that the user could only rely on the download function to view the log, the log viewer itself would create little value. Provide pause/resume feature could avoid this problem.
3. As for the Download function, is there any evidence that downloads all logs are more frequently used than downloads per container? I found Openshift has changed its design from supporting both download methods to only supporting downloading per container. It seems that the decision to support downloading per container may be based on its wider range of use cases and benefits. With regard to this point, I have[ another variation](https://www.figma.com/proto/rJjOtl3nxpW6BAcl0MZr4y/Pipeline-log-viewer?page-id=62%3A1895&type=design&node-id=82-1574&viewport=723%2C-184%2C0.24&t=S2A8z2MwR9c9Ffjg-1&scaling=min-zoom&starting-point-node-id=82%3A1574&mode=design) for the toolbar which displays the two download functions in a different hierarchy.

**Comment by andrewballantyne** [comment]
I think we have enough that we can start working on this. So removing the labels that block it from popping up on todo work for the team.

**Comment by kywalker-rh** [comment]
This is great @yih-wang! Thanks for working on it!

@kaedward Do you have an opinion on whether the button should be called "Resume" or "Refresh"? (See second and third screens of the mock-up)

**Comment by yih-wang** [comment]
> We can definitely start with container, the only drawback I can think of is, if they really want to download all, they have to manually go through the effort of selecting each and clicking download. I think the use-case of that would be just wanting to leave our UI and view it in a log tool of their own; I would imagine post-execution rather than during.

@andrewballantyne I think it makes sense to keep both download options (download logs of current container&all containers) in the drawer. My question was only about how to provide a better download experience from the ux perspective - whether to include both download options in a dropdown or prioritize the more frequently used one as the main action (which will save users a click each time they want to download). I understand your point about treating all containers as a unit since they work together as a task. Based on that, I would suggest:
1) Changing the action in the alert from downloading the log of the current container to downloading the logs of all containers. This provides a more intuitive default action for users.
2) Regarding the download dropdown above the log viewer, we can stick to [the previous mockup](https://www.figma.com/proto/rJjOtl3nxpW6BAcl0MZr4y/Pipeline-log-viewer?page-id=62%3A1895&type=design&node-id=82-1161&viewport=723%2C-184%2C0.24&t=S2A8z2MwR9c9Ffjg-1&scaling=min-zoom&starting-point-node-id=82%3A1161&mode=design) until we gather enough evidence (maybe through user interviews).

One more thing I missed in my initial mock is the error handling. When the log viewer cannot retrive the log as expected (e.g. due to network issue), we should provide a notification to inform users. Here is [an adding screen](https://www.figma.com/proto/rJjOtl3nxpW6BAcl0MZr4y/Pipeline-log-viewer?page-id=62%3A1895&type=design&node-id=141-2183&viewport=-1335%2C-2825%2C0.45&t=Pmf7rv5ysEZgGCnS-1&scaling=min-zoom&starting-point-node-id=141%3A2183&mode=design) to address this. What I'm not sure is whether user will encounter a blank page (fetch nothing) or they can still see the content from the last fetch in this case - this distinction determines whether they should see the expandable alert which indicates only partial content is contained.

A last question (hope it is) I have is regarding the scenario where the latest fetch retrieves nothing (no new logs generated in the past 3 seconds). Will the log viewer just show a blank window? I think it will be great if we can implement a condition like 'if there are no new logs, the previous log is not discarded'.

**Comment by andrewballantyne** [comment]
@yih-wang Thanks for the error handling situation... definitely something we should support.

> A last question (hope it is) I have is regarding the scenario where the latest fetch retrieves nothing (no new logs generated in the past 3 seconds). Will the log viewer just show a blank window? I think it will be great if we can implement a condition like 'if there are no new logs, the previous log is not discarded'.

It should always support pulling the log information irrelevant of the duration we fetch or how much we fetch. If there are logs there, we should get it. So on re-poll, it'll just not change (which is fine imo).

However, this does prompt an interesting question. What if there was _never_ logs? A container may not (doesn't have to) report anything. It _likely_ will... but I think you can get a situation were there are just _no logs_ for this container at this time. We should try to separate our "loading" state from our "here is the result" state. Do you have some guidance on what we want to show? Maybe we just show the initial state as a spinner & then if it's empty, there is just no logs? Or do we want to say "no logs" in some fashion?

**Comment by kaedward** [comment]
@kywalker-rh Not sure if this is an option, but could we show only the icon, and then have a tooltip? "Pause refreshing log" and "Resume refreshing log" ideally... I guess my answer to your question is "both", but that's probably too long for a static message here :-) If tooltip isn't an option, would "pause refreshing" and "resume refreshing" fit?

**Comment by yih-wang** [comment]
> However, this does prompt an interesting question. What if there was never logs? A container may not (doesn't have to) report anything. It likely will... but I think you can get a situation were there are just no logs for this container at this time. We should try to separate our "loading" state from our "here is the result" state. Do you have some guidance on what we want to show? Maybe we just show the initial state as a spinner & then if it's empty, there is just no logs? Or do we want to say "no logs" in some fashion?

@andrewballantyne That is a good point. This reminds me that the initial design lacks a system to indicate different status of the log. So I reconsider all the status that the log viewer should have and list them as following:
1. Loading status: this status will be displayed when the user first enters this tab. The log viewer may take a longer time to initially fetch a bulk of log lines from the server. (Not sure whether this could be differentiate from the Refreshing status though)
2. Streaming status: the browser fetches the latest log from the server successfully. The user is allowed to pause the next refreshing to take a closer look at the log.
3. Pause status: the fetch is pausing in this status. The log viewer will only continue fetching after user resume refreshing.
4. Refreshing status: appear every 3 seconds, or follow a Resume action. Users may not notice this status if the network speed is fast.
5. End status: there are no more new logs. So the user won't see any actions/status in the toolbar.

Based on these status, here is a demo that shows how a full cycle of the log looks like:

![full-cycle](https://github.com/opendatahub-io/odh-dashboard/assets/23331885/b23ec090-76ee-4919-a1a5-0736985256bb)

And here is a demo that addresses the use case you mentioned above - when there are no logs at all.

![no-log](https://github.com/opendatahub-io/odh-dashboard/assets/23331885/3929a67a-b98a-4909-b187-7e2b0ebe1ca6)

**Comment by kywalker-rh** [comment]
@kaedward I think the way @yih-wang shows it in the demo works great (which looks like it matches your suggestion). If the drawer this is in is too small to fit the text we could then just show the icon, but I like how this is done.

Thanks for adding the demo @yih-wang!

**Comment by kaedward** [comment]
Looks great to me, too - thanks @yih-wang! The text doesn't take up as much room as I thought it might.

**Comment by yih-wang** [comment]
I have compiled all the designs for the first pass into a single link to make it more convenient as a reference for implementation. Check this link: [Final design for the first pass](https://www.figma.com/proto/rJjOtl3nxpW6BAcl0MZr4y/Pipeline-log-viewer?page-id=62%3A1895&type=design&node-id=260-2386&viewport=664%2C-1095%2C0.23&t=80ExnogduAcwD3kx-1&scaling=min-zoom&starting-point-node-id=260%3A2386&mode=design).

@kywalker-rh @andrewballantyne I'm considering how to track the updates and additional features we may want to add in the future. Due to time limitations, some features haven't been included (e.g. downloading logs from S3 after the run ends, checking the previous version of running pod logs in case of a crash loop, etc.). Should we capture these ideas in this thread or somewhere else to ensure we address them in future iterations?

**Comment by ykoyfman** [comment]
Nice to see that logs will be accessible in the UI, that will be very helpful. Regarding:

> Tail the log file by 500 lines every 3 seconds -- don't cache any information, just show the current selected container
On each fetch we will toss the previous data away and just show the latest information (so the last - up to - 500 lines)

Any possibility to save all log data and allow users to scroll back to beginning of the output?  Thanks

**Comment by andrewballantyne** [comment]
> Nice to see that logs will be accessible in the UI, that will be very helpful. Regarding:
> 
> > Tail the log file by 500 lines every 3 seconds -- don't cache any information, just show the current selected container
> > On each fetch we will toss the previous data away and just show the latest information (so the last - up to - 500 lines)
> 
> Any possibility to save all log data and allow users to scroll back to beginning of the output? Thanks

I think given our current infrastructure @ykoyfman we are leaning into the download feature for the first pass. We cannot do that live (as the pod is running) due to the lack of WebSockets and not really having a clear understanding of where the logs started per tail or if we missed lines.

**Comment by andrewballantyne** [comment]
> @kywalker-rh @andrewballantyne I'm considering how to track the updates and additional features we may want to add in the future. Due to time limitations, some features haven't been included (e.g. downloading logs from S3 after the run ends, checking the previous version of running pod logs in case of a crash loop, etc.). Should we capture these ideas in this thread or somewhere else to ensure we address them in future iterations?

@yih-wang I recommend you guys log something on your side and track suggested improvements. Since they are not technical and we don't have requirements for that, until we know more about the need of that, there isn't much for Dev to track.

When we support WebSockets, this and other functionalities will be on Dev to support. If anyone plays around with the logs and notes a desired improvement, they can log something on our end and we will work with UX to see the viability. 

At least this is how we have done it up to this point... let me know if you think there is a better way.

**Comment by shalberd** [comment]
@andrewballantyne WebSockets or gRPC? I guess Websockets to to the patternfly library used, right?

**Comment by andrewballantyne** [comment]
> It's not possible to directly call a gRPC service from a browser. gRPC uses HTTP/2 features, and no browser provides the level of control required over web requests to support a gRPC client.
> * https://learn.microsoft.com/en-us/aspnet/core/grpc/browser?view=aspnetcore-7.0

I don't think gRPC is available to us in the browser 🤔 There seems to be middle ground items inside that link -- but I don't think this is really a discussion for anything we have planned today.

Websockets are needed to allow the server to push to the client. We are not looking for the speed or the frequency of gRPC at this time. I'll admit my knowledge of gRPC is the odd article here or there that I have read. I'll look into your link when I get a chance and see if there is an architectural change here for us to embrace. 

Also, small nit, Patterfly is a component library and doesn't prescribe anything about data layers. Websockets is the expected choice because K8s & our SDK (that we use to talk to K8s) have that functionality already built in. 

Final note, our websocket issue:
- https://github.com/opendatahub-io/odh-dashboard/issues/1138

Feel free to pitch alternatives to Websockets, but I don't really want this issue to change course. Neither will get in before we need to support logs (as that time as already come and gone and we need this functionality sooner rather than later).

**Comment by shalberd** [comment]
> small nit, Patterfly is a component library

yeah, I confused that. With patternfly library, I did mean your SDK.

> We are not looking for the speed or the frequency of gRPC at this time

totally ok, as I mentioned, it was more of an idea. Plus, as you correctly say, it'd be dependent on how to K8S APIs to e.g. logging serve up the data, whether that is even already gRPC serving or not. I think gRPC will be more of a topic for streaming-type data, which logs could be, in future in K8S
